PluginManagerThatSupportMultiLoader报错Didn't find class "android.content.ContextParams"

有两个单独的plugin1.zip  plugin2.zip   ,通过PluginManagerThatSupportMultiLoader用同一个进程加载。
流程1：先打开plugin1.zip，再打开plugin2.zip ，没问题
流程2：如果是先打开plugin2.zip，后打开plugin1.zip，就会报错

报错信息如下，这个问题应该如何排查？应该不是PluginManagerThatSupportMultiLoader用的有问题吧？不然流程1应该也会报错的吧
这是不是个BUG @shifujun 

11-30 06:47:29.742 18303 18303 E AndroidRuntime: java.lang.LinkageError: While checking class com.tencent.shadow.sample.plugin.runtime.PluginDefaultProxyActivity method android.content.Context com.tencent.shadow.core.runtime.container.GeneratedHostActivityDelegator.createContext(android.content.ContextParams) signature against interface com.tencent.shadow.core.runtime.container.GeneratedHostActivityDelegator: Failed to resolve arg 0 type android.content.ContextParams with com.tencent.shadow.dynamic.host.MultiDynamicContainer$ContainerClassLoader (declaration of 'com.tencent.shadow.sample.plugin.runtime.PluginDefaultProxyActivity' appears in /data/user/0/com.tencent.shadow.sample.host/files/ShadowPluginManager/UnpackedPlugin/test-dynamic-manager/817abc99b50019e57e0251301dbfdb05/plugin-debug.zip/sample-runtime-debug.apk)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at java.lang.Class.newInstance(Native Method)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.AppComponentFactory.instantiateActivity(AppComponentFactory.java:95)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.Instrumentation.newActivity(Instrumentation.java:1251)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3399)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3643)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:140)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:100)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2229)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:238)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7798)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:512)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:995)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/content/ContextParams;
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	... 15 more
**11-30 06:47:29.742 18303 18303 E AndroidRuntime: **Caused by: java.lang.ClassNotFoundException: Didn't find class "android.content.ContextParams"** on path: DexPathList[[zip file** "/data/user/0/com.tencent.shadow.sample.host/files/ShadowPluginManager/UnpackedPlugin/test-dynamic-manager/af1f71a4ccfdf01abec1c34b9d0784f6/plugin.zip/core-runtime-release.apk"],nativeLibraryDirectories=[/system/lib64, /system/product/lib64]]
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:196)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
11-30 06:47:29.742 18303 18303 E AndroidRuntime: 	... 15 more

--- a/projects/sdk/dynamic/dynamic-host-multi-loader-ext/src/main/java/com/tencent/shadow/dynamic/host/MultiDynamicContainer.java
+++ b/projects/sdk/dynamic/dynamic-host-multi-loader-ext/src/main/java/com/tencent/shadow/dynamic/host/MultiDynamicContainer.java
@@ -114,5 +114,23 @@ public class MultiDynamicContainer {
             this.containerKey = containerKey;
             this.apkFilePath = installedApk.apkFilePath;
         }
+
+        @Override
+        protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+            Class<?> c = findLoadedClass(name);
+            if (c == null) {
+                // 对于当前apk中的类不进行双亲委派查找
+                // 因为这个ClassLoader中的类比较特殊，Activity等壳子接口的方法上存在当前环境可能不存在的类，
+                // 比如ContextParams这种API 31新引入的类。如果采用双亲委派，PluginContainerActivity等基类
+                // 可能跨ClassLoader加载，会触发HasSameSignatureWithDifferentClassLoaders的校验，
+                // 进而加载所有方法签名上的类型，从而可能在低版本机器上报找不到类的错误。
+                try {
+                    c = findClass(name);
+                } catch (ClassNotFoundException ignored) {
+                    c = super.loadClass(name, resolve);
+                }
+            }
+            return c;
+        }
     }
 }

